app/dao/services_dao.py

```diff
@@ -194,16 +197,41 @@ def dao_fetch_service_by_inbound_number(number):
 
 
 def dao_fetch_service_by_id_with_api_keys(service_id, only_active=False):
-    query = Service.query.filter_by(
-        id=service_id
-    ).options(
-        joinedload('api_keys')
-    )
-
-    if only_active:
-        query = query.filter(Service.active)
+    @contextmanager
+    def get_session(engine):
+        session = scoped_session(sessionmaker(bind=engine))
+        try:
+            yield session
+        finally:
+            session.close()
+
+    reader = db.engines['read-db']
+    with get_session(reader) as session:
+        query = session.query(Service).filter_by(
+            id=service_id
+        ).options(
+            joinedload('api_keys')
+        )
 
-    return query.one()
+        if only_active:
+            query = query.filter(Service.active)
+
+        try:
+            result = query.one()
+            # instead of returning the whole model attached to read-db engine
+            # extract needed properties and return object that can be
+            # serialized for caching
+            return AuthenticatedServiceInfo(result)
+        except AuthenticatedServiceInfoException as err:
+            current_app.logger.error("Could not find service with ID %s", service_id)
+            raise NoResultFound(err)
+        except NoResultFound:
+            # we handle this failure in the parent
+            current_app.logger.error("Could not find service with ID %s", service_id)
+            raise
+        except Exception:
+            # we handle this failure in the parent
+            raise
```

Comment:
Correct, but I wanted to be explicit about what is actually going on because initially it wasn't. Everyone who looked at the code with me asked that question including you. I believe it's a good practice to do these so developers don't wonder why errors are not handled and if code is reused somewhere they'd know that parent needs to handle the error.